{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "7f05b8ba",
   "metadata": {},
   "source": [
    "### This is a program to download the gene features with Patric ID programmatically"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "0c4516d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATRIC ID</th>\n",
       "      <th>Isolation Source</th>\n",
       "      <th>Isolation Date</th>\n",
       "      <th>Serotype</th>\n",
       "      <th>US State of Isolation</th>\n",
       "      <th>SRA Run Accession</th>\n",
       "      <th>Antibiotic</th>\n",
       "      <th>Laboratory-derived MIC</th>\n",
       "      <th>Predicted MIC</th>\n",
       "      <th>Training set genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>590.11954</td>\n",
       "      <td>Hogs</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>Derby</td>\n",
       "      <td>NC</td>\n",
       "      <td>SRR3295894</td>\n",
       "      <td>AUG</td>\n",
       "      <td>&lt;=1</td>\n",
       "      <td>2.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>590.11954</td>\n",
       "      <td>Hogs</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>Derby</td>\n",
       "      <td>NC</td>\n",
       "      <td>SRR3295894</td>\n",
       "      <td>AMP</td>\n",
       "      <td>&lt;=1</td>\n",
       "      <td>2.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>590.11954</td>\n",
       "      <td>Hogs</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>Derby</td>\n",
       "      <td>NC</td>\n",
       "      <td>SRR3295894</td>\n",
       "      <td>AXO</td>\n",
       "      <td>&lt;=0.25</td>\n",
       "      <td>0.25</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>590.11954</td>\n",
       "      <td>Hogs</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>Derby</td>\n",
       "      <td>NC</td>\n",
       "      <td>SRR3295894</td>\n",
       "      <td>AZI</td>\n",
       "      <td>4</td>\n",
       "      <td>4.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>590.11954</td>\n",
       "      <td>Hogs</td>\n",
       "      <td>2014.0</td>\n",
       "      <td>Derby</td>\n",
       "      <td>NC</td>\n",
       "      <td>SRR3295894</td>\n",
       "      <td>CHL</td>\n",
       "      <td>8</td>\n",
       "      <td>8.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69108</th>\n",
       "      <td>590.17856</td>\n",
       "      <td>Ground Turkey</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Senftenberg</td>\n",
       "      <td>NM</td>\n",
       "      <td>SRR3664859</td>\n",
       "      <td>KAN</td>\n",
       "      <td>&gt;64</td>\n",
       "      <td>128.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69109</th>\n",
       "      <td>590.17856</td>\n",
       "      <td>Ground Turkey</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Senftenberg</td>\n",
       "      <td>NM</td>\n",
       "      <td>SRR3664859</td>\n",
       "      <td>NAL</td>\n",
       "      <td>4</td>\n",
       "      <td>4.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69110</th>\n",
       "      <td>590.17856</td>\n",
       "      <td>Ground Turkey</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Senftenberg</td>\n",
       "      <td>NM</td>\n",
       "      <td>SRR3664859</td>\n",
       "      <td>STR</td>\n",
       "      <td>&gt;64</td>\n",
       "      <td>128.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69111</th>\n",
       "      <td>590.17856</td>\n",
       "      <td>Ground Turkey</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Senftenberg</td>\n",
       "      <td>NM</td>\n",
       "      <td>SRR3664859</td>\n",
       "      <td>TET</td>\n",
       "      <td>&lt;=4</td>\n",
       "      <td>16.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69112</th>\n",
       "      <td>590.17856</td>\n",
       "      <td>Ground Turkey</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Senftenberg</td>\n",
       "      <td>NM</td>\n",
       "      <td>SRR3664859</td>\n",
       "      <td>TIO</td>\n",
       "      <td>1</td>\n",
       "      <td>1.00</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69113 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       PATRIC ID Isolation Source  Isolation Date     Serotype  \\\n",
       "0      590.11954             Hogs          2014.0        Derby   \n",
       "1      590.11954             Hogs          2014.0        Derby   \n",
       "2      590.11954             Hogs          2014.0        Derby   \n",
       "3      590.11954             Hogs          2014.0        Derby   \n",
       "4      590.11954             Hogs          2014.0        Derby   \n",
       "...          ...              ...             ...          ...   \n",
       "69108  590.17856    Ground Turkey             NaN  Senftenberg   \n",
       "69109  590.17856    Ground Turkey             NaN  Senftenberg   \n",
       "69110  590.17856    Ground Turkey             NaN  Senftenberg   \n",
       "69111  590.17856    Ground Turkey             NaN  Senftenberg   \n",
       "69112  590.17856    Ground Turkey             NaN  Senftenberg   \n",
       "\n",
       "      US State of Isolation SRA Run Accession Antibiotic  \\\n",
       "0                        NC        SRR3295894        AUG   \n",
       "1                        NC        SRR3295894        AMP   \n",
       "2                        NC        SRR3295894        AXO   \n",
       "3                        NC        SRR3295894        AZI   \n",
       "4                        NC        SRR3295894        CHL   \n",
       "...                     ...               ...        ...   \n",
       "69108                    NM        SRR3664859        KAN   \n",
       "69109                    NM        SRR3664859        NAL   \n",
       "69110                    NM        SRR3664859        STR   \n",
       "69111                    NM        SRR3664859        TET   \n",
       "69112                    NM        SRR3664859        TIO   \n",
       "\n",
       "      Laboratory-derived MIC  Predicted MIC Training set genome  \n",
       "0                        <=1           2.00                 Yes  \n",
       "1                        <=1           2.00                 Yes  \n",
       "2                     <=0.25           0.25                 Yes  \n",
       "3                          4           4.00                 Yes  \n",
       "4                          8           8.00                 Yes  \n",
       "...                      ...            ...                 ...  \n",
       "69108                    >64         128.00                 Yes  \n",
       "69109                      4           4.00                 Yes  \n",
       "69110                    >64         128.00                 Yes  \n",
       "69111                    <=4          16.00                 Yes  \n",
       "69112                      1           1.00                 Yes  \n",
       "\n",
       "[69113 rows x 10 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read the data\n",
    "import pandas as pd\n",
    "\n",
    "df = pd.read_excel(\"a4ed1601ccb774eaf3b9c5f44b1a6bc3_JCM.01260-18-s0002.xlsx\", skiprows=1)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "61413319",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5278"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Find out the list of non repeating patrid IDs\n",
    "pid_list = list(set(df[\"PATRIC ID\"]))\n",
    "len(pid_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1d3a2d1",
   "metadata": {},
   "source": [
    "#### Start download"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "974e49fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: cannot create directory ‘Features’: File exists\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "590_12285 downloaded, count: 1\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import os.path\n",
    "\n",
    "# Make a directory named \"Features in the root directory\"\n",
    "os.system(\"mkdir Features\")\n",
    "count = 0\n",
    "\n",
    "# Loop through the list of unique Patrid IDs and download the sequence with p3 api\n",
    "for pid in pid_list:\n",
    "    file_name = \"_\".join(str(pid).split(\".\"))\n",
    "    filename_full = \"Features/features_{}.fasta\".format(file_name)\n",
    "    \n",
    "    os.system(\"p3-genome-fasta --feature {} >{}\".format(pid, filename_full))\n",
    "    \n",
    "    #Check if the downloaded file is empty\n",
    "    if os.path.getsize(filename_full) > 0:\n",
    "        count += 1\n",
    "    else:\n",
    "        os.system(\"rm {}\".format(filename_full))\n",
    "    \n",
    "    # Count can be varied\n",
    "    if count > 500:\n",
    "        break\n",
    "    print(f\"{file_name} downloaded, count: {count}\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
